### ISSUE TITLE
[Enhancement] Improved program window scaling via super-sampled integer nearest neighbor

### ISSUE THREAD

**Issue by NintendoManiac64** [Arisotura/melonDS#1100]
The current "screen filtering" option seems to simply change the scaling algorithm from nearest neighbor to bilinear.

So my proposal is doing what I believe mGBA does for its "bilinear (pixelated)" scaling mode on the 3DS - doing integer nearest neighbor scaling to whatever would be the next largest ratio is that's bigger than the current window size, and then downscaling to the program window via bilinear.

For example, on a 1360x768 display with the DS screens arranged horizontally with the program window maximized, 1360px is 2.65625x that of the 512px of both DS screens side-by-side.  That is...not a particularly high ratio, meaning non-integer nearest neighbor looks pretty bad on anything that isn't an already blocky or pixelated game.

With my proposal, you'd instead scale that 512px to the next highest integer ratio which, in the case of a 1360px-wide display, would be 1536px wide - an exact 3x scale.  From there you would then downscale back to the display's 1360px width via a simple bilinear or the like.

(now to clarify this doesn't have to be exclusive to having the screens in a landscape orientation or even with the two screens at the same size; this was just simpler for me to explain my proposal using the situation I have found myself in)

Here's a visual example of what I mean, all scaled to 1360px wide:

**Nearest neighbor** (various diagonal slopes are ugly)
![nearest_opt](https://user-images.githubusercontent.com/11896003/117871087-861ac900-b28c-11eb-83f7-06486e496750.png)

**Bilinear** (blurry, 'nuff said)
![bilinear_opt](https://user-images.githubusercontent.com/11896003/117871098-8915b980-b28c-11eb-81f1-701fb2837e53.png)

**Super-sampled integer nearest neighbor**
![supersampled_opt](https://user-images.githubusercontent.com/11896003/117871109-8d41d700-b28c-11eb-834a-9beb53a1c6fe.png)

**Comment by nadiaholmquist** [comment]
I'm very in favor of adding this, nearest prescale + bilinear is definitely the way to go for this because as you show, it retains the crispness of the image without causing visibly uneven pixel sizes. I might try and implement this.

**Comment by poudink** [comment]
I'd like it to be optional. I like my scaling unfiltered.

**Comment by nadiaholmquist** [comment]
Of course, the screen filtering option would be replaced with a submenu (or box in the video settings?) with options for nearest, bilinear and this new option, "sharp" or some other friendly name so people who like the current options can keep those.

**Comment by NintendoManiac64** [comment]
If the nearest neighbor scaling gets revamped, then perhaps also a setting for emulating the pixel grid could also be implemented?  By its nature it'd probably only work correctly when integer scaling is used.

Basically, at sufficiently high ratios (e.g. 500%), it can actually be more visually accurate to surround each of the DS's pixels with a black boarder to simulate the look of an LCD in a manner that's kind of like the LCD equivalent of scanlines (and I guess it could similarly be configurable with regards to how thick you want the aforementioned pixel grid - on a 4k+ resolution display you'd probably need the grid to be two pixels thick rather than just one)

Example image at 500% with single-pixel width grid (you'll definitely need to zoom it in to full size to see what I mean):
![image](https://user-images.githubusercontent.com/11896003/119317490-ae7ccd00-bc67-11eb-9e85-3896af293e7e.png)

**Comment by poudink** [comment]
I get what you mean, but it seems like a pretty poor way to emulate the effect, IMO.

**Comment by memmam** [comment]
I was actually about to suggest this, except instead of scaling to the next integer scale and downscaling, you scale to the integer factor below that and then bilinear scale UP the rest of the way. This is the method used on the RetroTINK 5X and 4K, and produces a VERY clean image. At some scales, the fact that it's bilinear scaled at all is almost imperceptible.

I'd be curious to compare the two methods and see what gives a better result.

**Comment by NintendoManiac64** [comment]
It's my impression that what people call "sharp bilinear" nowadays is exactly what I describe, going from the next largest integer factor and then scaling down.

I suggested doing it this way because, when I did testing, this way consistently gave the best results.

I imagine the reason why the RetroTINK does it the other way is because they're limited by maximum resolution and can't deal with going above whatever their maximum is due to being a hardware-based solution.

**Comment by memmam** [comment]
As someone who was involved in the testing and documentation of the RetroTINK-4K, I can tell you that the maximum practical resolution constraint is high enough that I don't think that was a consideration, with the device able to hit a 18x scale from 240p (resulting in a 4K window of an 8K image) without issue (this would be used for e.g. determining proper phase when creating a profile).

Out of curiosity, I actually tested our two methods, and the result looks virtually identical between the two. See for yourself:

![5x](https://github.com/melonDS-emu/melonDS/assets/9062595/e8370548-f509-48c4-b1e3-b703da300412)

![6x](https://github.com/melonDS-emu/melonDS/assets/9062595/6aa02f25-f624-4d2e-b09d-c537061986cc)

**Comment by NintendoManiac64** [comment]
Use a smaller scale factor—the images in my examples are only around 2.6x which amplifies any differences.

Once you get to resolutions as high 8x+ the differences becomes indistinguishable; I mean, when you're turing 1 pixel into 8 or 9, that's almost to the point that you could even get away with plain standard non-integer nearest neighbor!

**Comment by memmam** [comment]
The reason I scaled to 5.625x is because it covers a common use case, scaling a single DS screen to 1080p.

Here's the same test using 720p as a target:

![3x](https://github.com/melonDS-emu/melonDS/assets/9062595/46d58e3b-2cdf-4086-9745-36befdb251a5)

![4x](https://github.com/melonDS-emu/melonDS/assets/9062595/137926a9-6fed-4d13-934c-84c9c72d9dc6)

Your method looks sharper under this test, and additionally that if you're scaling both DS screens (384 pixels tall) to 720p, you get a scale factor of 1.875x, where downscaling from 2x makes more sense. Additionally, the RetroTINK use case is based on the assumption you're scaling to at least 4x nearest neighbor to start, which mitigates the effect some.

Additionally, upon further reflection I may be misremembering and it may be nearest neighbor scaling to the *closest* integer factor, and then scaling up or down as appropriate. I'll need to check with Mike.
